src/core/types.ml

```diff
@@ -0,0 +1,66 @@
+open Gospel
+
+module Mutability = struct
+  let known_tysymbol ~driver (ts : Ttypes.tysymbol) =
+    Option.fold ~none:Translated.Unknown
+      ~some:(fun (t : Translated.type_) -> t.mutable_)
+      (Drv.get_type ts driver)
+
+  let tysymbol ~driver (ts : Ttypes.tysymbol) =
+    max
+      (if
+       Ttypes.ts_equal ts (Drv.get_ts driver [ "Gospelstdlib"; "array" ])
+       || Ttypes.ts_equal ts (Drv.get_ts driver [ "Gospelstdlib"; "ref" ])
+      then Translated.Mutable
+      else Translated.Immutable)
+      (known_tysymbol ~driver ts)
+
+  let rec ty ~driver (t : Ttypes.ty) =
+    match t.ty_node with
+    | Tyvar _ -> Translated.Unknown
+    | Tyapp (ts, tyl) ->
+        List.fold_left
+          (fun acc t -> max acc (ty ~driver t))
+          (tysymbol ~driver ts) tyl
+
+  let lsymbol ~driver (ls : Tterm.lsymbol) =
+    Option.fold ~none:Translated.Unknown ~some:(ty ~driver) ls.ls_value
+    |> List.fold_right (fun t acc -> max acc (ty ~driver t)) ls.ls_args
```

Comment:
I think you could `fold_left` instead to get the tail-recursiveness (not that it matter that much with these small lists...).

Reply:
It was for the pipe.